### PR TITLE
Allow etags to be disabled with config option

### PR DESCRIPTION
**Pull request by ooHmartY**
This PR allows etag generation to be disabled via a config option. 

The goal is to allow developers to decide when to send an etag header depending on the cache strategy being used.

This can also help save some CPU ticks by eliminating calls to crypto functions during SSR.

### CODE REVIEW

**Comment by rauchg** [comment]
This looks very good. Couple questions:
- Should we look into making etag generation less expensive? Etags seem super valuable to save bandwidth costs. Seem like a tradeoff between compute / end-user performance here.
- The option should be more explicit perhaps `generateEtags`

**Comment by rauchg** [comment]
It would be nice to have tests for this as well, since it should be fairly easy to do so.

**Comment by ooHmartY** [comment]
1. agreed i believe this can be done with by making the etag generation async. It also depends on the cache strategy in our particular use case etags are not used because of how we use CDN cache.
2. Easy change and it makes the option more explicit 👍

**Comment by rauchg** [comment]
Even if you use a CDN cache, the CDN can also avoid the re-transfer when performing an invalidation. The CDN is just another client, so ETag would still be desirable?

**Comment by arunoda** [comment]
@ooHmartY 

> agreed i believe this can be done with by making the etag generation async

I don't think even we do it we can get any CPU improvements. Since whether we do it async or not, we need to spend the same CPU cycles.
(Unless there's a some sort of workers for generatic etags. Anyway, that's an overkill)

File: readme.md [hunk 1]
```diff
@@ -724,6 +724,7 @@ The `next` API is as follows:
 
 Supported options:
 - `dev` (`bool`) whether to launch Next.js in dev mode - default `false`
+- `etag` (`bool`) whether to append etag header to HTML responses - default `true`
```
Comment (arunoda) [inline readme.md]: Usually, we don't accept configuration from here. We do it via our config file `next.config.js`. Could you use that instead?

And use a verb like `disableEtags`

**Review by arunoda — CHANGES_REQUESTED** [review]

File: readme.md (hunk 1 above)
Comment (rauchg) [reply readme.md]: We went for `generateEtags: false` @arunoda. See above

**Comment by ooHmartY** [comment]
@rauchg & @arunoda -- apologies for the late reply and thank you for your feedback... with this PR I'm advocating for having the ability to disable etags and save some CPU ticks in the process. 

In the past I've moved etag generation out of the application to a reverse proxy that's optimized for etag generation and caching to avoid hitting the application altogether.

The latest commit adds test & updates the config option to `generateEtags`.

File: server/index.js [hunk 2]
```diff
@@ -29,7 +29,7 @@ const blockedPages = {
 }
 
 export default class Server {
-  constructor ({ dir = '.', dev = false, staticMarkup = false, quiet = false, conf = null } = {}) {
+  constructor ({ dir = '.', dev = false, generateEtags = true, staticMarkup = false, quiet = false, conf = null } = {}) {
```
Comment (timneutkens) [inline server/index.js]: Lets move the `generateEtags` to `server/config.js` https://github.com/zeit/next.js/blob/canary/server/config.js#L5

You can then use it like `this.nextConfig.generateEtags`

And define `generateEtags` in next.config.js.

**Comment by ooHmartY** [comment]
@rauchg , @arunoda & @timneutkens -- I've updated this with latest canary. Please let me know if there are any additional changes needed.

**Comment by timneutkens** [comment]
@ooHmartY I've updated the PR to use renderOpts instead of passing the nextConfig through. @arunoda can you do a final review 🙏

**Review by arunoda — APPROVED** [review]
These changes looks great to me.

**Pull request merged**